src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py

```diff
@@ -0,0 +1,428 @@
+
+    params = {y[0]: y[1].strip('"') for y in
+              (x.strip().split('=', 2) for x in tokens[1].split(','))}
+    if 'realm' not in params or 'service' not in params:
+        accumulated_errors |= _error_handler(Errors.CONNECTIVITY_AAD_LOGIN_ERROR, "AAD access", ignore_errors)
+        return accumulated_errors # Can't continue if registry is not AAD-reachable
+
+    # Ensure tokens can be obtained
+    authurl = urlparse(params['realm'])
+    authhost = urlunparse((authurl[0], authurl[1], '/oauth2/exchange', '', '', ''))
+
+    from azure.cli.core._profile import Profile
+    profile = Profile(cli_ctx=cmd.cli_ctx)
+    creds, _, tenant = profile.get_raw_token()
+
+    headers = {'Content-Type': 'application/x-www-form-urlencoded'}
+    content = {
+        'grant_type': 'access_token',
+        'service': params['service'],
+        'tenant': tenant,
+        'access_token': creds[1]
+    }
+
+    response = requests.post(authhost, urlencode(content), headers=headers)
+
+    if response.status_code not in [200]:
+        accumulated_errors |= _error_handler(Errors.CONNECTIVITY_REFRESH_TOKEN_ERROR,
+                                             "Obtain refresh token for registry {}".format(login_server),
+                                             ignore_errors)
+        return accumulated_errors # We can't continue if the refresh token is not received
+
+    print_pass("Obtain refresh token for registry {}".format(login_server))
+
+    from json import loads
+
+    refresh_token = loads(response.content.decode("utf-8"))["refresh_token"]
+    authhost = urlunparse((authurl[0], authurl[1], '/oauth2/token', '', '', ''))
+    content = {
+        'grant_type': 'refresh_token',
+        'service': login_server,
+        'scope': 'registry:catalog:*',
+        'refresh_token': refresh_token
+    }
+
+    response = requests.post(authhost, urlencode(content), headers=headers)
+
+    if response.status_code not in [200]:
+        accumulated_errors |= _error_handler(Errors.CONNECTIVITY_ACCESS_TOKEN_ERROR,
+                                             "Obtain access token for registry {}".format(login_server),
+                                             ignore_errors)
+    else:
+        print_pass("Obtain access token for registry {}".format(login_server))
+
+    return accumulated_errors
+
+
+def _check_health_connectivity(cmd, registry_name, ignore_errors):
+    accumulated_errors = 0
+
+    if registry_name is None:
+        print_warning("Registry name must be provided to check connectivity.")
+        return accumulated_errors
+
+    try:
+        registry, _ = get_registry_by_name(cmd.cli_ctx, registry_name)
+        login_server = registry.login_server
+    except CLIError:
+        from ._docker_utils import get_login_server_suffix
+        suffix = get_login_server_suffix(cmd.cli_ctx)
+        if suffix is None:
+            suffix = '.azurecr.io'
```

Comment:
this is incorrect for national cloud. Again, please try to reuse existing code, feel free to refactor/improve them.

Reply:
Yes, I know. We try to retrieve the suffix according to the current cloud using `suffix = get_login_server_suffix(cmd.cli_ctx)`. In case we can't retrieve, I think that we should use the most common one.

Reply:
I wouldn't make that assumption. If you couldn't retrieve, you should just error out. But again, this duplicates a bunch of existing code, so please refactor.